### PR TITLE
	Preserve the serializer type when fragment caching

### PR DESCRIPTION
**Pull request by bdmac**
We were not previously cloning the type setting into the dynamically
generated cached/non-cached serializers for a given fragment-cached
serializer. This led to the type generated for JsonApi having the wrong
value when fragment caching is enabled by adding either :except or :only
options to cache.

This pulls the type setting from the fragment-cached serializer forward
onto the dynamic caching classes so it is preserved in the output.

### CODE REVIEW

File: lib/active_model/serializer/adapter/cached_serializer.rb [hunk 1]
```diff
@@ -24,7 +24,7 @@ def cached?
         end
 
         def fragment_cached?
-          @klass._cache_only && !@klass._cache_except || !@klass._cache_only && @klass._cache_except
+          @klass._cache && (@klass._cache_only && !@klass._cache_except || !@klass._cache_only && @klass._cache_except)
```
Comment (bf4) [inline lib/active_model/serializer/adapter/cached_serializer.rb]: this can be it's own PR. would you mind? https://github.com/CheckMateIO/active_model_serializers/commit/e9e2b5712df0eda12399bac31a5d126b8b040082

**Comment by bf4** [comment]
Awesome, thanks! :+1: some comments above

File: test/adapter/fragment_cache_test.rb [hunk 2]
```diff
@@ -31,8 +31,13 @@ def test_fragment_fetch_with_namespaced_object
           }
           assert_equal(@spam_hash.fetch, expected_result)
         end
+
+        def test_fragment_fetch_with_type_override
+          type_serializer = TypedRoleSerializer.new(Role.new(name: 'Another Author'))
+          type_hash       = ActiveModel::Serializer::Adapter::JsonApi.new(type_serializer).serializable_hash
+          assert_equal(type_hash[:data][:type], TypedRoleSerializer._type)
```
Comment (bf4) [inline test/adapter/fragment_cache_test.rb]: better to use the serializable test helper (below may contain typos)

```ruby
serialization = serializable(Role.new(name: 'Another Author'), serializer: TypedRoleSerializer, adapter: :json_api).serializable_hash
assert_equal(TypedRoleSerializer._type, serialization.fetch(:data).fetch(:type))
```

File: test/fixtures/poro.rb [hunk 3]
```diff
@@ -222,6 +222,18 @@ def maker
...
+  type 'my-roles'
+  cache only: [:name], skip_digest: true
+  attributes :id, :name, :description, :slug
+
+  def slug
+    "#{object.name}-#{object.id}"
+  end
+
+  belongs_to :author
+end
```
Comment (bf4) [inline test/fixtures/poro.rb]: would you mind putting this in the test file?  We're try to no longer add files to poro unless they are for general usage

File: test/fixtures/poro.rb [hunk 4]
```diff
@@ -222,6 +222,18 @@ def maker
...
 end
 
+TypedRoleSerializer = Class.new(ActiveModel::Serializer) do
+  type 'my-roles'
+  cache only: [:name], skip_digest: true
+  attributes :id, :name, :description, :slug
+
+  def slug
+    "#{object.name}-#{object.id}"
+  end
```
Comment (bf4) [inline test/fixtures/poro.rb]: noting that in `rc4` slug won't be call here. you'd have to do `attribute :slug { "#{object.name}-#{object.id}" }`.  can be removed from the poro here, though, since it's not being used

**Comment by bdmac** [comment]
@bf4 comments addressed. Will open a secondary PR with the fix for `fragment_cached?`

**Comment by bf4** [comment]
Rebase, merged, and changelog added in https://github.com/rails-api/active_model_serializers/commit/527c2aed98a9b8687726b691620334f4c4e0bb9e

**Comment by bf4** [comment]
Closed by https://github.com/rails-api/active_model_serializers/commit/527c2aed98a9b8687726b691620334f4c4e0bb9e

**Comment by bf4** [comment]
Closed by a40998273d1eb6594e1344cae41603f9a5966386

**Comment by bf4** [comment]
This was merged, but GitHub didn't pick it up. I'm not sure why
